frontend/catalyst/pennylane_extensions.py

```diff
@@ -1457,7 +1605,7 @@ def _decorator(true_fn: Callable):
     return _decorator
 
 
-def for_loop(lower_bound, upper_bound, step):
+def for_loop(lower_bound, upper_bound, step, preserve_dimensions: bool = False):
```

Comment:
I see this option as a consequence of the following facts: (1) In loop body programs, we need results to match arguments (2) We only do one-pass tracing (of Python program representing loop body). All we have is one particular set of arguments and we need to determine one particular set of results. This might be not enough to emit the correct program, so we need additional information which we can get either from options (or by doing second-pass of the tracing which I assume we don't want). So, by this switch I plan to cover the two equally-important cases so I see it as compromise. The cases are: to preserve all shared dimension variables in loops and to treat all dimension variables as distinct ones, even if they happened to be shared in a particular 0-iteration inputs that we have.

Reply:
Following the discussion from our meeting I think we settled on marking the flag experimental (`experimental_` prefix) and setting the default behaviour to `True`.

Don't forget to also documentation for this option.

Reply:
👍

Reply:
Done.